A1_perception:
  node:
    period_ms: 10
    climb_condition:
      pitch_alpha: 0.001
      enable_climbing_threshold: -1.70
      disable_climbing_threshold: -1.70
      timeout_ms: 1800
    inputs:
      robot_position:
        topic: "/amcl_pose"
        type: "PoseWithCovarianceStamped"
      one_d_tof:
        topic: "/sensor_to_pointcloud/tof/mono"
        type: "PointCloud2"
      lidar_pc_baselink:
        topic: "/scan_merged_pointcloud"
        type: "PointCloud2"
      multi_tof_left_row1:
        topic: "/sensor_to_pointcloud/tof/multi/left/row_1"
        type: "PointCloud2"
      multi_tof_left_row2:
        topic: "/sensor_to_pointcloud/tof/multi/left/row_2"
        type: "PointCloud2"
      multi_tof_left_row3:
        topic: "/sensor_to_pointcloud/tof/multi/left/row_3"
        type: "PointCloud2"
      multi_tof_left_row4:
        topic: "/sensor_to_pointcloud/tof/multi/left/row_4"
        type: "PointCloud2"
      multi_tof_right_row1:
        topic: "/sensor_to_pointcloud/tof/multi/right/row_1"
        type: "PointCloud2"
      multi_tof_right_row2:
        topic: "/sensor_to_pointcloud/tof/multi/right/row_2"
        type: "PointCloud2"
      multi_tof_right_row3:
        topic: "/sensor_to_pointcloud/tof/multi/right/row_3"
        type: "PointCloud2"
      multi_tof_right_row4:
        topic: "/sensor_to_pointcloud/tof/multi/right/row_4"
        type: "PointCloud2"
      collision:
        topic: "/sensor_to_pointcloud/collision"
        type: "PointCloud2"
        
      camera_points:
        topic: "/sensor_to_pointcloud/camera_object"
        type: "PointCloud2"
      imu:
        topic: "/imu_data"
        type: "sensor_msgs/msg/Imu"

    layers:
      one_d_tof:
        inputs:
        - "one_d_tof"
        filter:
          compose:
            filters:
              timeout:
                timeout_milliseconds: 60000
              lidar_dist_check:
                min_range: 0.0
                max_range: 0.2
              one_d_roi:
                use_inside: true
                polygon: [ [0.41, -0.5], [0.41, 0.5], [0.18,0.5], [0.18,-0.50]]
              one_d_tof:
                use_stop: true
                stop_and_back_dist: 0.33
                line_length: 0.10
                resolution: 0.05
      camera_points:
        inputs:
        - "camera_points"
        filter:
          compose:
            filters:
              timeout:
                timeout_milliseconds: 60000
              sector_roi:
                use_inside: true
                min_range: 0.3
                max_range: 5.0
                min_angle: -180
                max_angle: 180
              density:
                max_count: 3
                radius: 0.2
      multi_tof_12:
        inputs:
        - "multi_tof_left_row1"
        - "multi_tof_left_row2"
        - "multi_tof_right_row1"
        - "multi_tof_right_row2"
        filter:
          compose:
            filters:
              lidar_dist_check:
                min_range: 0.0
                max_range: 0.15
              timeout:
                timeout_milliseconds: 15000
              sector_roi:
                use_inside: true
                min_range: 1.01
                max_range: 2.5
                min_angle: -160
                max_angle: 160
              density:
                max_count: 3
                radius: 0.3
      multi_tof_12_realtime:
        inputs:
        - "multi_tof_left_row1"
        - "multi_tof_left_row2"
        - "multi_tof_right_row1"
        - "multi_tof_right_row2"
        filter:
          compose:
            filters:
              lidar_dist_check:
                min_range: 0.0
                max_range: 0.15
              timeout:
                timeout_milliseconds: 100
              sector_roi:
                use_inside: true
                min_range: 0.3
                max_range: 1.0
                min_angle: -50.3
                max_angle: 50.3
              density:
                max_count: 3
                radius: 0.1

      # warning_low_obstacle:
      #   inputs:
      #   - "multi_tof_right_row3"
      #   - "multi_tof_left_row3"
      #   filter:
      #     compose:
      #       filters:
      #         timeout:
      #           timeout_milliseconds: 100
      #         sector_roi:
      #           use_inside: true
      #           min_range: 0.55
      #           max_range: 0.8
      #           min_angle: -50.3
      #           max_angle: 50.3

      
      # multi_tof_left_34_fusion:
      #   inputs:
      #   - "multi_tof_left_row3"
      #   filter:
      #     compose:
      #       filters:
      #         lidar_dist_check:
      #           min_range: 0.0
      #           max_range: 0.1
      #         low_obstacle:
      #           dist_diff: 0.1
      #           inputs:
      #           - multi_tof_left_row4
      
      # multi_tof_right_34_fusion:
      #   inputs:
      #   - "multi_tof_right_row3"
      #   filter:
      #     compose:
      #       filters:
      #         lidar_dist_check:
      #           min_range: 0.0
      #           max_range: 0.1
      #         low_obstacle:
      #           dist_diff: 0.1
      #           inputs:
      #           - multi_tof_right_row4

      multi_tof_4_left:
        inputs:
        - "multi_tof_left_row4"
        filter:
          compose:
            filters:
              lidar_dist_check:
                min_range: 0.0
                max_range: 0.20
              drop_off:
                min_range: 0.59 
                line_length: 0.05
                resolution: 0.05
                compare_dist_diff_max: 0.65
                inputs:
                - multi_tof_left_row3
                
      multi_tof_4_right:
        inputs:
        - "multi_tof_right_row4"
        filter:
          compose:
            filters:
              lidar_dist_check:
                min_range: 0.0
                max_range: 0.20
              drop_off:
                min_range: 0.59 
                line_length: 0.05
                resolution: 0.05
                compare_dist_diff_max: 0.65
                inputs:
                - multi_tof_right_row3
      collision:
        inputs:
        - "collision"
        filter:
          compose:
            filters:
              collision:
                line_length: 0.1
                resolution: 0.05
